packages/shared/comms/v4/BFFConnection.ts

```diff
@@ -0,0 +1,187 @@
+import { ILogger, createLogger } from 'shared/logger'
+import { Observable } from 'mz-observable'
+import { AuthIdentity, Authenticator } from '@dcl/crypto'
+import { createRpcClient, RpcClientPort, Transport } from '@dcl/rpc'
+import { WebSocketTransport } from '@dcl/rpc/dist/transports/WebSocket'
+import { loadService, RpcClientModule } from '@dcl/rpc/dist/codegen'
+import { BffAuthenticationServiceDefinition, WelcomePeerInformation } from './proto/bff/authentication-service.gen'
+import { CommsServiceDefinition } from './proto/bff/comms-service.gen'
+
+export declare type BFFConfig = {
+  getIdentity: () => AuthIdentity
+}
+
+export type TopicData = {
+  peerId: string
+  data: Uint8Array
+}
+
+export type TopicListener = {
+  subscriptionId: number
+}
+
+export class BFFConnection {
+  private logger: ILogger = createLogger('BFF: ')
+
+  public onDisconnectObservable = new Observable<void>()
+  public onTopicMessageObservable = new Observable<TopicData>()
+
+  private wsTransport: Transport | null = null
+
+  private sceneTopics = new Map<string, TopicListener>()
+
+  private commsService: RpcClientModule<CommsServiceDefinition, any> | null = null
+
+  constructor(private url: string, private config: BFFConfig) {}
+
+  async connect(): Promise<string> {
+    this.wsTransport = WebSocketTransport(new WebSocket(this.url, 'comms'))
+    this.wsTransport.on('close', async () => {
+      this.logger.log('transport closed')
+      this.disconnect()
+    })
+    this.wsTransport.on('error', async () => {
+      this.logger.log('transport closed')
+      this.disconnect()
+    })
+    const rpcClient = await createRpcClient(this.wsTransport)
+    const port = await rpcClient.createPort('kernel')
+    const peerId = await this.authenticate(port)
+
+    this.commsService = loadService(port, CommsServiceDefinition)
+    this.logger.log('Connected')
+
+    return peerId
+  }
+
+  public async addPeerTopicListener(
+    topic: string,
+    handler: (data: Uint8Array, peerId: string) => Promise<void>
+  ): Promise<TopicListener> {
+    if (!this.commsService) {
+      throw new Error('BFF is not connected')
+    }
+
+    const subscription = await this.commsService.subscribeToPeerMessages({ topic })
+
+    ;(async (commsService) => {
+      for await (const { payload, sender } of commsService.getPeerMessages(subscription)) {
+        await handler(payload, sender)
+      }
+    })(this.commsService).catch((err) => {
+      this.logger.error(`Peer topic handler error: ${err.toString()}`)
+      this.disconnect()
+    })
```

Comment:
Prefer named functions over auto-invoked functions
```suggestion
    async function getAsyncMessages(commsService) {
      for await (const { payload, sender } of commsService.getPeerMessages(subscription)) {
        await handler(payload, sender)
      }
    }
    
    getAsyncMessages(this.commsService).catch((err) => {
      this.logger.error(`Peer topic handler error: ${err.toString()}`)
      this.disconnect()
    })
```